src/net.cpp

```diff
@@ -2577,44 +2583,73 @@ bool StopNode()
 
 void NetCleanup()
 {
-    LOCK(cs_vNodes);
-
-    // Close sockets
-    for (CNode *pnode : vNodes)
+    // clean up some globals (to help leak detection)
     {
-        if (pnode->hSocket != INVALID_SOCKET)
-            CloseSocket(pnode->hSocket);
+        LOCK(cs_vNodes);
+
+        // Close sockets
+        for (CNode *pnode : vNodes)
+        {
+            // Since we are quitting, disconnect abruptly from the node rather than finishing up our conversation
+            // with it.
+            pnode->vRecvMsg.clear();
+            pnode->ssSend.clear();
+            pnode->nSendSize = 0;
+            // Now close communications with the other node
+            pnode->CloseSocketDisconnect();
+        }
+        for (ListenSocket &hListenSocket : vhListenSocket)
+        {
+            if (hListenSocket.socket != INVALID_SOCKET)
+                if (!CloseSocket(hListenSocket.socket))
+                    LOG(NET, "CloseSocket(hListenSocket) failed with error %s\n",
+                        NetworkErrorString(WSAGetLastError()));
+        }
     }
-    for (ListenSocket &hListenSocket : vhListenSocket)
+
+    // Try to let nodes be cleaned up for a while, but ultimately give up because we are shutting down.
+    for (int iters = 0; iters < 20; iters++)
     {
-        if (hListenSocket.socket != INVALID_SOCKET)
-            if (!CloseSocket(hListenSocket.socket))
-                LOG(NET, "CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
+        CleanupDisconnectedNodes();
+        {
+            LOCK(cs_vNodes);
+            if ((vNodes.size() == 0) && (vNodesDisconnected.size() == 0))
+                break; // every node is properly disconnected
+        }
+        MilliSleep(100); // Give other threads a chance to finish up using the node.
     }
 
-    // clean up some globals (to help leak detection)
-    for (CNode *pnode : vNodes)
-        delete pnode;
-    for (CNode *pnode : vNodesDisconnected)
-        delete pnode;
-    vNodes.clear();
-    vNodesDisconnected.clear();
-    vhListenSocket.clear();
-    if (semOutbound)
-        delete semOutbound;
-    semOutbound = nullptr;
-    // BU: clean up the "-addnode" semaphore
-    if (semOutboundAddNode)
-        delete semOutboundAddNode;
-    semOutboundAddNode = nullptr;
-    if (pnodeLocalHost)
-        delete pnodeLocalHost;
-    pnodeLocalHost = nullptr;
+
+    {
+        LOCK(cs_vNodes);
+        if (!((vNodes.size() == 0) && (vNodesDisconnected.size() == 0)))
+        {
+            LOG(NET, "Some node objects were not properly cleaned up.\n");
+        }
+
+        // If the nodes were not properly shut down, remove them from the vNodes list now, so the vNode item
+        // is not leaked.
+        // The node memory itself will be leaked but since we are quitting this is not a big issue.
```

Comment:
is it possible to have the "still-working" threads somewhat don't let the main process quit?

Reply:
No.  In the for loop above, I give other threads only a few seconds after the cnodes have been disconnected to finish up and release their references.  However, if they do not properly do so, the loop times out and the system is shut down anyway (without deleting the cnode), with a log indicating that there was a cleanup problem.  I did it this way because its "benign" to shutdown even though we aren't fully cleaned up.  But it makes memory leak detection harder so its not ideal.

While working on this, I noticed that other threads were in fact not properly releasing cnode references, so I fixed those cases.  

I also made a change to that tells the script queue threads to quit themselves before the globals are cleaned up.  This absolutely ensures that the script queue threads are not accidentally using some globals after they are deconstructed as opposed to the prior way which relied on them being in their idle state. It also makes shutdown easier to debug since we don't have to page through lots of idle threads to find the one that is still busy.